Adding PagerDuty provider

heimweh opened the pull request:
This is my first contribution and my first time writing Go, 
and I will most likely need some handholding during this process, but here goes.

Due to the missing PagerDuty support I decided to take a look at writing a provider for it. 

It currently supports:
```hcl
provider "pagerduty" {}

resource "pagerduty_team" "foo" {
  name        = "Engineering"
  description = "All engineering"
}

resource "pagerduty_user" "foo" {
  name  = "Earline Greenholt"
  email = "125.greenholt.earline@graham.name"
  teams = ["${pagerduty_team.foo.id}"]
}

resource "pagerduty_escalation_policy" "foo" {
  name      = "Engineering Escalation Policy"
  num_loops = 2
  teams     = ["${pagerduty_team.foo.id}"]

  rule {
    escalation_delay_in_minutes = 10

    target {
      id = "${pagerduty_user.foo.id}"
    }
  }
}

resource "pagerduty_schedule" "foo" {
  name      = "foo"
  time_zone = "Europe/Berlin"

  layer {
    name                         = "Primary Schedule Layer"
    start                        = "2015-11-06T20:00:00-05:00"
    rotation_virtual_start       = "2015-11-06T20:00:00-05:00"
    rotation_turn_length_seconds = 86400
    users                        = ["${pagerduty_user.foo.id}"]

    restriction {
      type              = "daily_restriction"
      start_time_of_day = "08:00:00"
      duration_seconds  = 32400
    }
  }
}

resource "pagerduty_service" "foo" {
  name                    = "My Web App"
  auto_resolve_timeout    = 14400
  acknowledgement_timeout = 600
  escalation_policy       = "${pagerduty_escalation_policy.foo.id}"
}

data "pagerduty_vendor" "datadog" {
  name_regex = "Datadog"
}

resource "pagerduty_service_integration" "foo" {
  name    = "${data.pagerduty_vendor.datadog.name}"
  type    = "${data.pagerduty_vendor.datadog.type}"
  vendor  = "${data.pagerduty_vendor.datadog.id}"
  service = "${pagerduty_service.foo.id}"
}
```

I've been testing it a lot in the past couple of days and the basic functionality is there, but as you might notice my knowledge of Go is currently somewhat limited.

```bash
$ make testacc TEST=./builtin/providers/pagerduty/
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/24 14:21:53 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/pagerduty -v  -timeout 120m
=== RUN   TestAccPagerDutyVendor_Basic
--- PASS: TestAccPagerDutyVendor_Basic (9.57s)
=== RUN   TestAccPagerDutyEscalationPolicy_import
--- PASS: TestAccPagerDutyEscalationPolicy_import (7.86s)
=== RUN   TestAccPagerDutySchedule_import
--- PASS: TestAccPagerDutySchedule_import (9.37s)
=== RUN   TestAccPagerDutyService_import
--- PASS: TestAccPagerDutyService_import (9.74s)
=== RUN   TestAccPagerDutyTeam_import
--- PASS: TestAccPagerDutyTeam_import (2.71s)
=== RUN   TestAccPagerDutyUser_import
--- PASS: TestAccPagerDutyUser_import (4.30s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccPagerDutyEscalationPolicy_Basic
--- PASS: TestAccPagerDutyEscalationPolicy_Basic (9.74s)
=== RUN   TestAccPagerDutyEscalationPolicyWithTeams_Basic
--- PASS: TestAccPagerDutyEscalationPolicyWithTeams_Basic (11.60s)
=== RUN   TestAccPagerDutySchedule_Basic
--- PASS: TestAccPagerDutySchedule_Basic (8.98s)
=== RUN   TestAccPagerDutySchedule_Multi
--- PASS: TestAccPagerDutySchedule_Multi (7.15s)
=== RUN   TestAccPagerDutyServiceIntegration_Basic
--- PASS: TestAccPagerDutyServiceIntegration_Basic (26.88s)
=== RUN   TestAccPagerDutyService_Basic
--- PASS: TestAccPagerDutyService_Basic (11.29s)
=== RUN   TestAccPagerDutyTeam_Basic
--- PASS: TestAccPagerDutyTeam_Basic (3.64s)
=== RUN   TestAccPagerDutyUser_Basic
--- PASS: TestAccPagerDutyUser_Basic (5.73s)
=== RUN   TestAccPagerDutyUserWithTeams_Basic
--- PASS: TestAccPagerDutyUserWithTeams_Basic (12.67s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/pagerduty	141.251s
```

Any feedback is more than welcome and I would love to see some continuation/improvements of this provider.

kwilczynski commented:
@heimweh hi there! This is great, absolutely great.

Once you get the go-pagerduty sorted, consider vendring it - this will fix the build error. Also, for anything that consumes JSON from users, I would recommend setting up a `ValidateFunc` plus normalising JSON (compacting/minifying). This can be done using `normalizeJsonString` helper function.

heimweh commented:
Hi! :) thanks for the feedback, sounds good, I'll definitely look into this asap.

kwilczynski commented:
@heimweh Sir, you are a Gentleman and a Scholar. This is absolutely amazing!

heimweh commented:
Thanks @kwilczynski! really appreciate it :) 

I've made quite some changes lately and that changed the logic quite a bit, sorry about that.
Hopefully the resources look a bit more clearer now though. 

I'll try to refactor my code a bit, I feel like I made a bit of a mess in there when moving things around.

heimweh commented:
@kwilczynski 
Hope all is well, sorry for the delay on this one.

I've made quite some changes lately, hope I didn't make it worse, if so please just let me know 😅
 
The docs are still a bit outdated, will take a look at that as soon as possible. 
Tests are still passing though and `go-pagerduty` is up-to-date and seems stable for now after a couple of changes in the client. 

I was also thinking about adding new validations since (after the changes) I'm not using JSON anymore, but I'm not quite sure on how to proceed. 

I guess I could probably add some basic validation checks, like in `validatePagerDutyUserRole`. 
Any ideas? or should we let the API give feedback instead? :)

Have a great weekend!

heimweh commented:
Documentation should now be up-to-date 🙂

kwilczynski commented:
@heimweh Sir, this is an outstanding piece of work!

kwilczynski commented:
@heimweh hi! I had a look, and you are on fire! A lot of amazing changes, etc. Few files probably need a run of `gofmt` over them, and here and there I would change the way how you access a map in Go to make it idiomatic, but otherwise things look good!

Thank you so much! ❤️

kwilczynski commented on builtin/providers/pagerduty/resource_pagerduty_escalation_policy.go:
```diff
@@ -0,0 +1,163 @@
+	d.Set("name", e.Name)
+	d.Set("description", e.Description)
+	d.Set("num_loops", e.NumLoops)
+	d.Set("escalation_rule", flattenRules(e.EscalationRules))
```
We often check for an error when setting using a return value from a function, just in case (albeit, we don't do it for a primitive/basic types).

```go
if err := d.Set("escalation_rule", flattenRules(e.EscalationRules)); err != nil {
    return err
}
```

heimweh replied:
Should be fixed now :)

kwilczynski commented on builtin/providers/pagerduty/resource_pagerduty_service.go:
```diff
@@ -0,0 +1,148 @@
+		service.AcknowledgementTimeout = &acknowledgementTimeout
+	}
+
+	policy := &pagerduty.EscalationPolicy{}
```
You could just do:

```go
service.EscalationPolicy = &pagerduty.EscalationPolicy{
    ID:   d.Get("escalation_policy").(string),
    Type: "escalation_policy",
}
```

heimweh replied:
Pushed a fix now :)

heimweh commented on builtin/providers/pagerduty/resource_pagerduty_schedule.go:
```diff
@@ -0,0 +1,217 @@
+
+	d.Set("name", s.Name)
+	d.Set("description", s.Description)
+	d.Set("schedule_layer", flattenLayers(s.ScheduleLayers))
```
Should be fixed now :)

heimweh commented on builtin/providers/pagerduty/resource_pagerduty_user.go:
```diff
@@ -0,0 +1,236 @@
+	validRoles := []string{"admin", "limited_user", "owner", "read_only_user", "user"}
+	role := v.(string)
+
+	if !contains(validRoles, role) {
```
Thank you so much. Just pushed a fix for this. 
Re-used a similar function found in the `Github` provider and seems to work fine :) is that better?

heimweh commented:
@kwilczynski I changed from `TypeSet` to `TypeList` for `resource_pagerduty_schedule`. 
Also got rid of a bug when you have more than 1 schedule layer attached. Tests are passing.

heimweh commented:
@kwilczynski 

I cleaned up a bit, updated the documentation and added support to attach teams to escalation policies. 

I think in terms of `resources` this should be okay for this PR, unless someone's missing something :) 

I've now also tested importing (`real/active`) existing users, schedules and escalation policies and that worked out quite well so now I have them managed in Terraform.

stack72 commented:
Hi @heimweh 

Thanks for all the work here - just in preparation to getting this merged, what level of pagerduty plan is needed to run the entire set of tests for this provider?

Paul

heimweh commented:
Hi @stack72, to run the entire set of tests a `standard` plan is required, however I think the only resource that actually needs the `standard` plan is the `teams` resource :)

This is the output when running the tests using a `basic` plan.

```bash
$ make testacc TEST=./builtin/providers/pagerduty
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/24 15:02:30 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/pagerduty -v  -timeout 120m
=== RUN   TestAccPagerDutyVendor_Basic
--- PASS: TestAccPagerDutyVendor_Basic (11.42s)
=== RUN   TestAccPagerDutyEscalationPolicy_import
--- PASS: TestAccPagerDutyEscalationPolicy_import (9.29s)
=== RUN   TestAccPagerDutySchedule_import
--- PASS: TestAccPagerDutySchedule_import (8.84s)
=== RUN   TestAccPagerDutyService_import
--- PASS: TestAccPagerDutyService_import (12.74s)
=== RUN   TestAccPagerDutyTeam_import
--- FAIL: TestAccPagerDutyTeam_import (0.41s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* pagerduty_team.foo: Failed call API endpoint. HTTP response code: 402. Error: &{2014 Required abilities are unavailable [The teams account ability is required to access teams]}
=== RUN   TestAccPagerDutyUser_import
--- PASS: TestAccPagerDutyUser_import (8.07s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccPagerDutyEscalationPolicy_Basic
--- PASS: TestAccPagerDutyEscalationPolicy_Basic (11.65s)
=== RUN   TestAccPagerDutyEscalationPolicyWithTeams_Basic
--- FAIL: TestAccPagerDutyEscalationPolicyWithTeams_Basic (5.70s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* pagerduty_team.foo: Failed call API endpoint. HTTP response code: 402. Error: &{2014 Required abilities are unavailable [The teams account ability is required to access teams]}
=== RUN   TestAccPagerDutySchedule_Basic
--- PASS: TestAccPagerDutySchedule_Basic (12.02s)
=== RUN   TestAccPagerDutySchedule_Multi
--- PASS: TestAccPagerDutySchedule_Multi (9.37s)
=== RUN   TestAccPagerDutyServiceIntegration_Basic
--- PASS: TestAccPagerDutyServiceIntegration_Basic (32.07s)
=== RUN   TestAccPagerDutyService_Basic
--- PASS: TestAccPagerDutyService_Basic (15.92s)
=== RUN   TestAccPagerDutyTeam_Basic
--- FAIL: TestAccPagerDutyTeam_Basic (0.44s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* pagerduty_team.foo: Failed call API endpoint. HTTP response code: 402. Error: &{2014 Required abilities are unavailable [The teams account ability is required to access teams]}
=== RUN   TestAccPagerDutyUser_Basic
--- PASS: TestAccPagerDutyUser_Basic (7.03s)
=== RUN   TestAccPagerDutyUserWithTeams_Basic
--- FAIL: TestAccPagerDutyUserWithTeams_Basic (0.46s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* pagerduty_team.foo: Failed call API endpoint. HTTP response code: 402. Error: &{2014 Required abilities are unavailable [The teams account ability is required to access teams]}
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/pagerduty	145.436s
make: *** [testacc] Error 1
```

stack72 commented:
Thanks for the info - will get on with a final review now :)

heimweh commented:
Awesome, thank you so much :)

kwilczynski commented:
@heimweh @stack72 the only things from me would be potentially (just before we merge):

- Remove surplus `&schema.Schema` (since Go 1.7+ these are not needed)
- Introduce `errwrap.Wrapf` over `fmt.Errorf` in the provider code (tests are different kettle of fish)
- Use Go style way of querying maps

But, none of it is really super-essential to get it merged.

Having said that, @heimweh did a truly AMAZING job! I cannot express enough how outstanding this Pull Request really is!

heimweh commented:
@kwilczynski thanks for your feedback! :) removed the unnecessary use of &schema.Schema. Will definitely take a look at the rest.

stack72 commented:
thanks for the work here - this is a really solid looking PR! :)

I agree with the feedback that @kwilczynski has suggested - these can be a follow up PR so we can get on with getting this merged in and having the nightly tests run as expected :)

heimweh commented:
Thanks guys 😊  I really appreciate the feedback :) 

This community is just pure **awesomeness**, so happy to be able to contribute.
That sounds good 👍  would be awesome to do a follow up soon. 

Thanks again! :)

stack72 commented:
@heimweh this works fantastically :)

```
% make testacc TEST=./builtin/providers/pagerduty                                                                                          ✭
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/24 17:03:46 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/pagerduty -v  -timeout 120m
=== RUN   TestAccPagerDutyVendor_Basic
--- PASS: TestAccPagerDutyVendor_Basic (11.17s)
=== RUN   TestAccPagerDutyEscalationPolicy_import
--- PASS: TestAccPagerDutyEscalationPolicy_import (8.22s)
=== RUN   TestAccPagerDutySchedule_import
--- PASS: TestAccPagerDutySchedule_import (8.83s)
=== RUN   TestAccPagerDutyService_import
--- PASS: TestAccPagerDutyService_import (13.31s)
=== RUN   TestAccPagerDutyTeam_import
--- PASS: TestAccPagerDutyTeam_import (2.96s)
=== RUN   TestAccPagerDutyUser_import
--- PASS: TestAccPagerDutyUser_import (4.66s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccPagerDutyEscalationPolicy_Basic
--- PASS: TestAccPagerDutyEscalationPolicy_Basic (11.44s)
=== RUN   TestAccPagerDutyEscalationPolicyWithTeams_Basic
--- PASS: TestAccPagerDutyEscalationPolicyWithTeams_Basic (14.46s)
=== RUN   TestAccPagerDutySchedule_Basic
--- PASS: TestAccPagerDutySchedule_Basic (9.50s)
=== RUN   TestAccPagerDutySchedule_Multi
--- PASS: TestAccPagerDutySchedule_Multi (8.57s)
=== RUN   TestAccPagerDutyServiceIntegration_Basic
--- PASS: TestAccPagerDutyServiceIntegration_Basic (30.01s)
=== RUN   TestAccPagerDutyService_Basic
--- PASS: TestAccPagerDutyService_Basic (13.53s)
=== RUN   TestAccPagerDutyTeam_Basic
--- PASS: TestAccPagerDutyTeam_Basic (4.82s)
=== RUN   TestAccPagerDutyUser_Basic
--- PASS: TestAccPagerDutyUser_Basic (8.30s)
=== RUN   TestAccPagerDutyUserWithTeams_Basic
--- PASS: TestAccPagerDutyUserWithTeams_Basic (17.40s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/pagerduty	167.225s
```

heimweh commented:
Awesome! thank you so much @stack72 ! :)
